osh/split.py

```diff
@@ -317,3 +321,128 @@ def Split(self, s, allow_escape):
+            if self.glob_escape:
+                frag = glob_.GlobEscapeBackslash(frag)
+            self.frags.append(frag)
+            del self.char_buff[:]
+
+    def _GenerateWord(self):
+        # type: () -> None
+        self._FlushCharBuff()
+        self.args.append(''.join(self.frags))
+        del self.frags[:]
+
+        if self.max_split >= 0 and len(self.white_buff) >= 1:
+            self.char_buff.extend(self.white_buff)
+            del self.white_buff[:]
+
+    def PushLiteral(self, s):
+        # type: (str) -> None
+        """
+        Args:
+          s: word fragment that should be literally added
+        """
+        if self.state == state_i.DE_White1:
+            self._GenerateWord()
+        else:
+            self._FlushCharBuff()
+        self.frags.append(s)
+        self.state = state_i.Black
+
+    def PushFragment(self, s):
+        # type: (str) -> None
+        """
+        Args:
+          s: word fragment to split
+        """
+        ifs_space = self.ifs_space
+        ifs_other = self.ifs_other
+        allow_escape = self.allow_escape
+        max_split = self.max_split
+        n = len(s)
+
+        for i in xrange(n):
+            byte = mylib.ByteAt(s, i)
+
+            if self.state == state_i.Backslash:
+                pass
+
+            elif max_split >= 0 and len(self.args) == max_split + 1:
+                # When max_split is reached, the processing is modified.
+                if allow_escape and byte == pyos.BACKSLASH_CH:
+                    self.state = state_i.Backslash
+                    continue
+                elif mylib.ByteInSet(byte, ifs_space):
+                    if self.state == state_i.Start:
+                        self.char_buff.append(byte)
+                        continue
+
+            elif allow_escape and byte == pyos.BACKSLASH_CH:
+                if self.state == state_i.DE_White1:
+                    self._GenerateWord()
+                self.state = state_i.Backslash
+                continue
+            elif mylib.ByteInSet(byte, ifs_space):
+                if self.state != state_i.Start:
+                    if len(self.args) == max_split:
+                        self.white_buff.append(byte)
+                    self.state = state_i.DE_White1
+                continue
+            elif mylib.ByteInSet(byte, ifs_other):
+                if len(self.args) == max_split:
+                    self.white_buff.append(byte)
+                self._GenerateWord()
+                self.state = state_i.Start
+                continue
+
+            if self.state == state_i.DE_White1:
+                self._GenerateWord()
+            self.char_buff.append(byte)
+            self.state = state_i.Black
+
+    def PushTerminator(self):
+        # type: () -> List[str]
+        if self.state in (state_i.DE_White1, state_i.Black):
+            self._GenerateWord()
+            if self.max_split >= 0 and len(self.args) == self.max_split + 2:
```

Comment:
Hm why do we need to "fix this up" here?  Is it possible to avoid calling _GenerateWord() in some cases?  That is what appends to self.args

But then it looks like we MAY have TOO MANY args after calling _GenerateWord() ?  So we have to do some joining

If it's not possible to avoid, can you write a comment about why that is?

Reply:
This is because it is the Bash behavior. When `read` requests $n$ words, and there are $n+m$ words, the extraction rule for the first $n$ words and the rest $m$ words are different. We need to read the $n$th word and the extra words in a different rule, where we need to keep them separately.